src/ring_theory/discriminant.lean

```diff
@@ -0,0 +1,221 @@
+  `σⱼ (b i)`, where `σⱼ : L →ₐ[K] E` is the embedding in an algebraically closed field `E`
+  corresponding to `j : ι` via a bijection `e : ι ≃ (L →ₐ[K] E)`.
+* `of_power_basis_eq_prod` : the discriminant of a power basis.
+
+## Implementation details
+Our definition works for any `A`-algebra `B`, but note that if `B` is not free as an `A`-module,
+then `trace A B = 0` by definition, so `discriminant A b = 0` for any `b`.
+-/
+
+universes u v w z
+
+open_locale matrix big_operators
+
+open matrix finite_dimensional fintype polynomial finset
+
+namespace algebra
+
+variables (A : Type u) {B : Type v} (C : Type z) {ι : Type w}
+variables [comm_ring A] [comm_ring B] [algebra A B] [comm_ring C] [algebra A C]
+
+section matrix
+
+/-- Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define
+`trace_matrix A ι b` as the matrix whose `(i j)`-th element is the trace of `b i * b j`. -/
+@[simp] noncomputable
+def trace_matrix (b : ι → B) : matrix ι ι A
+| i j := trace_form A B (b i) (b j)
+
+lemma trace_matrix_def (b : ι → B) : trace_matrix A b = λ i j, trace_form A B (b i) (b j) := rfl
+
+variable {A}
+
+lemma trace_matrix_of_matrix_vec_mul [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).vec_mul b) = Pᵀ ⬝ (trace_matrix A b) ⬝ P :=
+begin
+  ext α β,
+  rw [trace_matrix, vec_mul, dot_product, vec_mul, dot_product, mul_apply, bilin_form.sum_left,
+    fintype.sum_congr _ _ (λ (i : ι), @bilin_form.sum_right _ _ _ _ _ _ _ _
+    (b i * P.map (algebra_map A B) i α) (λ (y : ι), b y * P.map (algebra_map A B) y β)), sum_comm],
+  congr, ext x,
+  rw [mul_apply, sum_mul],
+  congr, ext y,
+  rw [map_apply, trace_form_apply, mul_comm (b y), ← smul_def],
+  simp only [id.smul_eq_mul, ring_hom.id_apply, map_apply, transpose_apply, linear_map.map_smulₛₗ,
+    trace_form_apply, algebra.smul_mul_assoc],
+  rw [mul_comm (b x), ← smul_def],
+  ring_nf,
+  simp,
+end
+
+lemma trace_matrix_of_matrix_mul_vec [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).mul_vec b) = P ⬝ (trace_matrix A b) ⬝ Pᵀ :=
+begin
+  refine add_equiv.injective transpose_add_equiv _,
+  rw [transpose_add_equiv_apply, transpose_add_equiv_apply, ← vec_mul_transpose,
+    ← transpose_map, trace_matrix_of_matrix_vec_mul, transpose_transpose, transpose_mul,
+    transpose_transpose, transpose_mul]
+end
+
+lemma trace_matrix_of_basis [decidable_eq ι] [fintype ι] (b : basis ι A B) :
+  trace_matrix A b = bilin_form.to_matrix b (trace_form A B) :=
+begin
+  ext i j,
+  rw [trace_matrix, trace_form_apply, trace_form_to_matrix]
+end
+
+variable (A)
+
+/-- `embeddings_matrix A C b : matrix ι (B →ₐ[A] C) C` is the matrix whose `(i, σ)` coefficient is
+  `σ (b i)`. It is mostly useful for fields when `fintype.card ι = finrank A B` and `C` is
+  algebraically closed. -/
+def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C := (λ i (σ : B →ₐ[A] C), σ (b i))
```

Comment:
```suggestion
@[simp] def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C
| i σ := σ (b i))
```
might be better, as that results in `rw embeddings_matrix` only working when it has indices, which prevents matrices decaying into functions in unexpected ways.